redash/query_runner/__init__.py

```diff
@@ -145,6 +147,93 @@ def _get_tables_stats(self, tables_dict):
+    response_error = "Endpoint returned unexpected status code"
+    requires_authentication = False
+    url_title = 'URL base path'
+    username_title = 'HTTP Basic Auth Username'
+    password_title = 'HTTP Basic Auth Password'
+
+    @classmethod
+    def configuration_schema(cls):
+        schema = {
+            'type': 'object',
+            'properties': {
+                'url': {
+                    'type': 'string',
+                    'title': cls.url_title,
+                },
+                'username': {
+                    'type': 'string',
+                    'title': cls.username_title,
+                },
+                'password': {
+                    'type': 'string',
+                    'title': cls.password_title,
+                },
+            },
+            'required': ['url'],
+            'secret': ['password']
+        }
+        if cls.requires_authentication:
+            schema['required'] += ['username', 'password']
+        return schema
+
+    def get_auth(self):
+        username = self.configuration.get('username')
+        password = self.configuration.get('password')
+        if username and password:
+            return (username, password)
+        if self.requires_authentication:
+            raise ValueError("Username and Password required")
+        else:
+            return None
+
+    def get_response(self, url, auth=None, **kwargs):
+        # Get authentication values if not given
+        if auth is None:
+            auth = self.get_auth()
+
+        # Then call requests to get the response from the given endpoint
+        # URL optionally, with the additional requests parameters.
+        error = None
+        response = None
+        try:
+            response = requests.get(url, auth=auth, **kwargs)
+            # Raise a requests HTTP exception with the appropriate reason
+            # for 4xx and 5xx response status codes which is later caught
+            # and passed back.
+            response.raise_for_status()
+
+            # Any other responses (e.g. 2xx and 3xx):
+            if response.status_code != 200:
+                error = '{} ({}).'.format(
+                    self.response_error,
+                    response.status_code,
+                )
+
+        except requests.HTTPError as exc:
+            logger.exception(exc)
+            error = (
+                "Failed to execute query. "
+                "Return Code: {} Reason: {}".format(
+                    response.status_code,
+                    response.text
+                )
+            )
+        except requests.RequestException as exc:
+            # Catch all other requests exceptions and return the error.
+            logger.exception(exc)
+            error = str(exc)
+        except Exception as exc:
+            # Catch any other exceptions, log it and reraise it.
+            logger.exception(exc)
+            raise sys.exc_info()[1], None, sys.exc_info()[2]
```

Comment:
We have this block of code in some places, but it's not really needed. Any unhandled exception is being taken care of by the execute query task.

Reply:
@arikfr I've removed those sections in https://github.com/getredash/redash/pull/2318/commits/8a4391ca436b15123caee9e78abb5d62dc8e103d (except the one in sqlite.py which does some extra unicode error handling).